synapse/storage/databases/main/receipts.py

```diff
@@ -541,14 +534,20 @@ def insert_linearized_receipt_txn(
         # have to compare orderings of existing receipts
         if stream_ordering is not None:
             sql = (
-                "SELECT stream_ordering, event_id FROM events"
-                " INNER JOIN receipts_linearized as r USING (event_id, room_id)"
-                " WHERE r.room_id = ? AND r.receipt_type = ? AND r.user_id = ?"
+                "SELECT e.stream_ordering, e.event_id, r.receipt_type FROM events AS e"
+                " INNER JOIN receipts_linearized AS r USING (event_id, room_id)"
+                " WHERE r.room_id = ? AND r.user_id = ?"
             )
-            txn.execute(sql, (room_id, receipt_type, user_id))
-
-            for so, eid in txn:
-                if int(so) >= stream_ordering:
+            txn.execute(sql, (room_id, user_id))
+
+            for so, eid, rt in txn:
+                if int(so) >= stream_ordering and (
+                    receipt_type == rt
+                    or (
+                        rt == ReceiptTypes.READ
+                        and receipt_type == ReceiptTypes.READ_PRIVATE
+                    )
```

Comment:
This was meant to prevent the read receipts from going "up" in the timeline to an older event. This happens when we have a read-receipt for event B and then receive a read receipt for event A (which is older than B) of a different type. We then just get the latest read-receipt that the server has received which is the read-receipt for event A.

This doesn't always work but should be fixable. Anyway, I am actually not sure what is the most optimal solution here. 

Should we ignore new read receipts for events for which we already have a read receipt of a similar type (the types that influence notif counts) or should we somehow try to handle this when getting the read receipts from the DB instead?

Reply:
I've pushed some more changes with the fix and added tests though I am still not exactly sure about the best solution here

Reply:
My thought was that we should simplify this logic here and always insert the latest read receipt / hidden read receipt into the room (ensuring separately they don't go "up" in a room), and then when we fetch the (read) receipt we can do some logic on them to see which is "later". We could do this in a couple of ways:

* By updating `get_receipts_for_user` to be `get_latest_receipt_for_user` and modifying it to search across multiple receipt types (it will require some custom SQL, but I think it will be fairly straightforward!)
* Renaming `get_receipts_for_user` to `get_read_receipt_for_user` and embedding the read vs. hidden read receipt logic into it.

In either way the logic should be pretty similar.